man/dgvalidator.1
.TH "DGVALIDATOR" "1" "4 May 2020" "dgvalidator" ""

.SH NAME
dgvalidator \- drumgizmo drumkit validator

.SH SYNOPSIS
\fBdgvalidator\fR [\fIOPTIONS\fR] <\fIFILE\fR>

.SH "DESCRIPTION"
.PP
\fBDgvalidator\fR is a simple validator tool for drumgizmo drumkit files.
.PP
Dgvalidator will try to load the drumkit pointed to by drumkitfile and
validate it. It validates both the xml in the drumkit and all instrument files
referred in it as well as the the semantics of the contents -
including checking for the existence and validity of all audiofiles
referred by the instruments.

.SH "OPTIONS"
.PD 0

.RE
\fB-n, --no-audio\fR
.RS 7
Skip checking audio file existence and samplerate.

.RE
\fB-v, --verbose\fR
.RS 7
Print more info during validation. Can be added multiple times to increase output verbosity.

.RE
\fB-q, --quiet\fR
.RS 7
Don't print any output, even on errors.

.RE
\fB-V, --version\fR
.RS 7
Print version and exit.

.RE
\fB-h, --help\fR
.RS 7
Print this message and exit.

.RE
\fB-D, --debug \fIDDD\fR
.RS 7
Enable debug messages on \fIDDD\fR see hugin documentation for details.

.RE
\fBdrumkitfile\fR
.RS 7
The drumkitfile to verify.

.RE
.SH "EXIT STATUS"
Exit status is 0 if no error are found and 1 if validation fails.

.RE
.SH "BUGS"
Report bugs to http://www.drumgizmo.org/wiki/doku.php?id=bugs.

.SH "ADDITIONAL INFORMATION"
For further information, visit the website http://www.drumgizmo.org.


ChangeLog
Version 0.9.18.1-hotfix
======================
Release date: October 23rd 2019
Contributors: deva

- Fixed bug in the file browser which could not be navigated if the
  default kit path were not set or pointing to an invalid path.

Special thanks to 2Late and Jason N. for reporting and helping debugging
the problem.

Version 0.9.18
==============
Release date: October 15th 2019
Contributors: chaot4, meka, deva

This release is primarily a bugfix release but a few new features also
managed to sneak in.

Highlights:
 - Sample selection algorithm now behaves a lot better when using a small
   sample set.
 - Error reporting has been drastically improved when loading drum-kits.
 - Sample normalization option has been added.

Also some bug fixes, mainly for the UI.

For a more in depth description of the changes and fixes, see the
changelog for 0.9.18:
 https://drumgizmo.org/wiki/doku.php?id=changelog:drumgizmo-0.9.18

Version 0.9.17
==============
Release date: July 15th 2019
Contributors: deva, chaot4, suhr, trebmuh

In this release we tried to jam in as many of our planned changes to the
xml as possible. There are still a few more to go but this release has
brought us a large step closer to our feature-complete xml format. Quite a
few other major features has also found their way into this release, which
will be much more visible (and audible) to the users.

Highlights:
 - Improved sample selection algorithm.
 - Clicky-kit feature.
 - Midnam support.
 - Directed choke functionality.
 - Drumkit metadata support.
 - Experimental Cocoa UI support for MacOSX.

Also some bug fixes and a lot of code refactoring (replaced eXpat with
pugixml for one).

For a detailed description of all the new shiny features, including
audio samples and lots of screenshots, see the changelog for 0.9.17:
 https://drumgizmo.org/wiki/doku.php?id=changelog:drumgizmo-0.9.17

Version 0.9.16
==============
Release date: August 12th 2018
Contributors: deva, chaot4, suhr, trebmuh

Resampler has been refactored.
It now depends strictly on zita-resampler and is no longer optional.

Bugfixes:
 - Fix latency reporting in VST.
 - Fix timing humanizer stops normal cymbal samples too early.
 - Samples are cut off when looping.
 - LV2: Don't export any symbols except the plugin entry point to avoid clashes with the DAW or other plugins.
 - VST: Don't export any symbols except the plugin entry point to avoid clashes with the DAW or other plugins.
 - Report latency from resampling.
 - Fix LV2 Buffer overflow.
 - Add message to execution of CLI without arguments that mentions that the GUI is only supported in the plugins.
 - Update French translation of manpage.
 - Fix crash or noise when changing jack buffersize while running.
 - Stuttering when resampling is active.

Version 0.9.15
==============
Release date: July 24th 2018
Contributors: deva, chaot4, meka, suhr, Muldjord, cglocke, Roman Garifullin, James Cowgill

New engine features:
 - New humanized timing feature.
 - Master bleed control.
 - Remove memory warning.
 - Replaced cppunit with new dgunit header-only test framework.

New CLI features:
 - OSS midi input support.
 - Finish and integrate getoptpp as a submodule.

UI:
 - Double click knobs to reset them to their default value.
 - Ignore spacebar key in UI (used by DAW and toggled the buttons).
 - Add cute little image in the inline display.

Bugfixes:
 - Fix X11 BadValue crash bug.
 - Fix black render when opened as LV2 on Ardour and FreeBSD.
 - Fix UI crash on underflow in resize()
 - Fix valgrind errors of GUI.
 - Fix crash if CLI is invoked without input and/or output engine. (deva)
 - Fix out of bounds drawing of frame (chaot4)

Version 0.9.14
==============
Note: LAC2017 release
Release date: May 14th 2017
Contributors: deva, chaot4, Muldjord, meka, Hans Petter Selasky
Special extra thanks with a cherry on top: Robin Gareus

Features:
 - Introduced new resampler module in ui

Bugfixes:
 - CLI
   - CLI exits right away, doesn't do anything
   - Make ''-i midifile -I midimap=~/...'' recognize the home directory. Replace "~" by the home directory in ./drumgizmo/input/midifile.cc
   - Properly get/set OSS buffers
   - Add ---no-resampling option

 - Engine
   - Fix midi note on/off mask
   - Streamer BUG: If there is a cache miss on one of the audiofiles (1 channel) in a sample that one sample will get delayed compared to the others
   - Streamer BUG: When nodata buffer is being reallocated we need to make sure that the old buffer is not being used by one of the cache ids
   - Resampler BUG: Fix corruption/crash when enabling/disabling resampler at run-time
   - Update Window build environment to a more recent mingw64
   - Remove obsolete old mingw32 thread hacks
   - Introduce clear signal, showing if we had an underrun because of diskstreaming
   - Fix crash on with low preload size
   - Initial status of drumkit is "Error" on first opening the plugin

 - UI
   - Scroll-wheel on tab buttons should iterate through the tabs
   - Refactor TextEdit. (use vector in preprocessing etc.)
   - Fix "unexpected easter egg"
   - Fix high CPU load of plugingui when shown in a host (All platforms/DAWs)
   - Return handle to the newly created window in ''PluginLV2::uiInstantiate'' in the ''widget'' pointer
   - Add bypass button for the resampler
   - Fix redraw on changing desktops

 - General
   - Don't put .a and .la files in lv2 folder on install

Version 0.9.13
==============
Release date: April 23rd 2017
Contributors: deva, chaot4, Muldjord, meka

Highlighted changes / fixes:
 - New ui
 - Diskstreaming
 - OSS support

All changes / fixes:
 - Fixed rcgen compilation with BSD make
 - Made ui and engine compile as internal libs instead of compiling all sources in each component
 - Implemented settings for controlling diskstreaming
 - Complete gui makeover
 - Made changes to UI to incorporate diskstreaming controls
   - Now generates resources at compile-time
   - Made window resize flicker-free on X11

drumgizmo/dgvalidator.cc
/* -*- Mode: c++ -*- */
/***************************************************************************
 *            dgvalidator.cc
 *
 *  Sun Jan 27 10:44:44 CET 2019
 *  Copyright 2019 Bent Bisballe Nyeng
 *  deva@aasimon.org
 ****************************************************************************/

/*
 *  This file is part of DrumGizmo.
 *
 *  DrumGizmo is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  DrumGizmo is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with DrumGizmo; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
 */
#include <dgxmlparser.h>
#include <path.h>
#include <domloader.h>
#include <random.h>
#include <settings.h>
#include <drumkit.h>
#include <iostream>
#include <string>
#include <hugin.hpp>
#include <getoptpp.hpp>
#include <sstream>
#include <climits>

#include <lodepng/lodepng.h>

#include <config.h>
#include <platform.h>

#if DG_PLATFORM != DG_PLATFORM_WINDOWS
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#endif

#include <image.h>

// Needed for Resource class
#include <resource_data.h>
const rc_data_t rc_data[] = {};

namespace
{

int verbosity{1};

void logger(LogLevel level, const std::string& message)
{
	switch(level)
	{
	case LogLevel::Info:
		if(verbosity < 3)
		{
			return;
		}
		break;
	case LogLevel::Warning:
		if(verbosity < 2)
		{
			return;
		}
		break;
	case LogLevel::Error:
		if(verbosity < 1)
		{
			return;
		}
		break;
	}

	switch(level)
	{
	case LogLevel::Info:
		std::cout << "[Info]";
		break;
	case LogLevel::Warning:
		std::cout << "[Warning]";
		break;
	case LogLevel::Error:
		std::cout << "[Error]";
		break;
	}
	std::cout << " " << message << std::endl;
}

std::string version()
{
	std::ostringstream output;
	output << "DGValidator v" << VERSION << std::endl;
	return output.str();
}

std::string copyright()
{
	std::ostringstream output;
	output << "Copyright (C) 2008-2020 DrumGizmo team - DrumGizmo.org.\n";
	output << "This is free software.  You may redistribute copies of it under the terms ";
	output << "of\n";
	output << "the GNU Lesser General Public License <http://www.gnu.org/licenses/gpl.html>.\n";
	output << "There is NO WARRANTY, to the extent permitted by law.\n";
	output << "\n";
	return output.str();
}

std::string usage(const std::string& name, bool brief = false)
{
	std::ostringstream output;
	output <<
		"Usage: " << name << " [options] <drumkitfile>\n";
	if(!brief)
	{
		output <<
			"\n"
			"Validates the xml and semantics of the drumkit file and prints any found"
			" errors to the console.\n"
			"Returns 0 on success or 1 if errors were found.\n"
			"\n";
	}
	return output.str();
}

bool pathIsFile(const std::string& path)
{
#if DG_PLATFORM == DG_PLATFORM_WINDOWS
	return (GetFileAttributesA(path.data()) & FILE_ATTRIBUTE_DIRECTORY) == 0;
#else
	struct stat s;
	if(stat(path.data(), &s) != 0)
	{
		return false; // error
	}

	return (s.st_mode & S_IFREG) != 0; // s.st_mode & S_IFDIR => dir
#endif
}

}

int main(int argc, char* argv[])
{
	bool no_audio{false};

	std::string hugin_filter;
	unsigned int hugin_flags = 0;
#ifndef DISABLE_HUGIN
	hugin_flags = HUG_FLAG_DEFAULT;
#endif /*DISABLE_HUGIN*/

	dg::Options opt;

	opt.add("no-audio", no_argument, 'n',
	        "Skip checking audio file existence and samplerate.",
	        [&]()
	        {
		        no_audio = true;
		        return 0;
	        });

	opt.add("verbose", no_argument, 'v',
	        "Print more info during validation. Can be added multiple times to"
	        " increase output verbosity.",
	        [&]()
	        {
		        ++verbosity;
		        return 0;
	        });

	opt.add("quiet", no_argument, 'q',
	        "Don't print any output, even on errors.",
	        [&]()
	        {
		        verbosity = 0;
		        return 0;
	        });

	opt.add("version", no_argument, 'V',
	        "Print version and exit.",
	        [&]()
	        {
		        std::cout << version();
		        exit(0);
		        return 0;
	        });

	opt.add("help", no_argument, 'h',
	        "Print this message and exit.",
	        [&]()
	        {
		        std::cout << usage(argv[0]);
		        std::cout << "Options:\n";
		        opt.help();
		        exit(0);
		        return 0;
	        });

#ifndef DISABLE_HUGIN
	opt.add("debug", required_argument, 'D',
	        "Enable debug messages on 'ddd' see hugin documentation for details.",
	        [&]()
	        {
		        hugin_flags |= HUG_FLAG_USE_FILTER;
		        hugin_filter = optarg;
		        return 0;
	        });
#endif /*DISABLE_HUGIN*/

	if(opt.process(argc, argv) != 0)
	{
		return 1;
	}

	if(opt.arguments().empty())
	{
		std::cerr << "Missing kitfile." << std::endl;
		std::cerr << usage(argv[0], true) << std::endl;
		return 1;
	}

	hug_status_t status = hug_init(hugin_flags, HUG_OPTION_FILTER,
	                               hugin_filter.data(), HUG_OPTION_END);
	if(status != HUG_STATUS_OK)
	{
		std::cerr << "Error: " << status << std::endl;
		return 1;
	}

	std::string edited_filename = opt.arguments()[0];
	DrumkitDOM drumkitdom;
	std::vector<InstrumentDOM> instrumentdoms;
	std::string path = getPath(edited_filename);
	bool parseerror = false;
	bool ret = parseDrumkitFile(edited_filename, drumkitdom, logger);
	if(!ret)
	{
		WARN(drumkitloader, "Drumkit file parser error: '%s'",
		     edited_filename.data());
	}

	parseerror |= !ret;

	for(const auto& ref : drumkitdom.instruments)
	{
		instrumentdoms.emplace_back();
		bool ret = parseInstrumentFile(path + "/" + ref.file, instrumentdoms.back(),
		                               logger);
		if(!ret)
		{
			WARN(drumkitloader, "Instrument file parser error: '%s'",
			     edited_filename.data());
		}

		parseerror |= !ret;
	}

	if(parseerror)
	{
		return 1;
	}

	Settings settings;
	Random rand;
	DrumKit kit;

	DOMLoader domloader(settings, rand);
	ret = domloader.loadDom(path, drumkitdom, instrumentdoms, kit, logger);
	if(!ret)
	{
		WARN(drumkitloader, "DOMLoader error");
		logger(LogLevel::Error, "Validator found errors.");
		return 1;
	}
	parseerror |= !ret;
	if(parseerror)
	{
		ERR(drumgizmo, "Drumkit parser failed: %s\n", edited_filename.c_str());
		logger(LogLevel::Error, "Validator found errors.");
		return 1;
	}

	if(no_audio == false)
	{
		// Verify all referred audiofiles
		for(const auto& instrument: kit.instruments)
		{
			for(auto& audiofile: instrument->audiofiles)
			{
				audiofile->load(logger, 1);
				if(!audiofile->isLoaded())
				{
					WARN(drumkitloader, "Instrument file load error: '%s'",
					     audiofile->filename.data());
					logger(LogLevel::Warning, "Error loading audio file '" +
					       audiofile->filename + "' in the '" + instrument->getName() +
					       "' instrument");
					parseerror = true;
				}
			}
		}
	}

	bool image_error{false};
	// Check drumkit images
	{
		if(!drumkitdom.metadata.image_map.empty() &&
		   drumkitdom.metadata.image.empty())
		{
			logger(LogLevel::Warning, "Found drumkit image_map but no image,"
			       " so image_map will not be usable.");
			image_error = true;
		}

		std::pair<std::size_t, std::size_t> image_size;
		if(!drumkitdom.metadata.image.empty())
		{
			// Check if the image file exists
			auto image = path + "/" + drumkitdom.metadata.image;
			logger(LogLevel::Info, "Found drumkit image '" + image + "'");
			if(!pathIsFile(image))
			{
				logger(LogLevel::Warning, "Image file does not exist.");
				image_error = true;
			}
			else
			{
				// Check if the image_map can be loaded (is a valid png file)
				GUI::Image img(image);
				if(!img.isValid())
				{
					logger(LogLevel::Warning, "Drumkit image, '" + image +
					       "', could not be loaded. Not a valid PNG image?");
					image_error = true;
				}
				else
				{
					image_size = { img.width(), img.height() };
					logger(LogLevel::Info, "Loaded image in resolution " +
					       std::to_string(image_size.first) + " x " +
					       std::to_string(image_size.second));
				}
			}
		}

		std::pair<std::size_t, std::size_t> image_map_size;
		if(!drumkitdom.metadata.image_map.empty())
		{
			// Check if the image_map file exists
			auto image_map = path + "/" + drumkitdom.metadata.image_map;
			logger(LogLevel::Info, "Found drumkit image_map '" + image_map + "'");
			if(!pathIsFile(image_map))
			{
				logger(LogLevel::Warning, "Image map file does not exist.");
				image_error = true;
			}
			else
			{
				// Check if the image_map can be loaded (is a valid png file)
				GUI::Image image(image_map);
				if(!image.isValid())
				{
					logger(LogLevel::Warning, "Drumkit image_map, '" + image_map +
					       "', could not be loaded. Not a valid PNG image?");
					image_error = true;
				}
				else
				{
					image_map_size = { image.width(), image.height() };
					logger(LogLevel::Info, "Loaded image_map in resolution " +
					       std::to_string(image_map_size.first) + " x " +
					       std::to_string(image_map_size.second));

					// Check if the click map colours can be found in the image_map image.
					for(const auto& clickmap : drumkitdom.metadata.clickmaps)
					{
						if(clickmap.colour.size() != 6)
						{
							logger(LogLevel::Warning,
							       "Clickmap colour field not the right length (should be 6).");
							image_error = true;
							continue;
						}

						try
						{
							auto hex_colour = std::stoul(clickmap.colour, nullptr, 16);
							float red   = (hex_colour >> 16 & 0xff) / 255.0f;
							float green = (hex_colour >>  8 & 0xff) / 255.0f;
							float blue  = (hex_colour >>  0 & 0xff) / 255.0f;
							GUI::Colour colour(red, green, blue);

							bool found{false};
							for(int y = 0; y < image.height() && !found; ++y)
							{
								for(int x = 0; x < image.width() && !found; ++x)
								{
									if(image.getPixel(x, y) == colour)
									{
										found = true;
									}
								}
							}

							if(!found)
							{
								logger(LogLevel::Warning,
								       "Clickmap colour '" + clickmap.colour +
								       "' not found in image_map.");
								image_error = true;
							}
						}
						catch(...)
						{
							// Not valid hex number
							logger(LogLevel::Warning,
							       "Clickmap colour not a valid hex colour.");
							image_error = true;
							continue;
						}

						// Check if the click map instruments exist.
						bool found{false};
						for(const auto& instrument: kit.instruments)
						{
							if(instrument->getName() == clickmap.instrument)
							{
								found = true;
							}
						}
						if(!found)
						{
							logger(LogLevel::Warning,
							       "Clickmap instrument '" + clickmap.instrument +
							       "' not found in drumkit.");
							image_error = true;
						}

					}
				}
			}
		}

		// Check if the image and the image_map have same resolutions
		if(image_size != image_map_size)
		{
			logger(LogLevel::Warning,
			       "Drumkit image and image_map does not have same resolution.");
			image_error = true;
		}

	}

	if(parseerror || image_error)
	{
		logger(LogLevel::Warning, "Validator found errors.");
		return 1;
	}

	logger(LogLevel::Info, "Validator finished without errors.");
	return 0;
}
